crates/matrix-sdk-sled/src/state_store.rs

```diff
@@ -507,6 +507,28 @@ impl SledStore {
 
                     for (room, events) in &changes.stripped_members {
                         for event in events.values() {
+                            let key = (room, &event.state_key);
+
+                            match event.content.membership {
+                                MembershipState::Join => {
+                                    joined.insert(
```

Comment:
I think this is problematic. This is now mixing up user ids that we pick out from the normal state with those we pick out from the stripped state.

This means that we could start making E2EE related decisions based on information coming from the stripped state. The list of joined and invited user ids is directly used to decide which members get the current room key.

Stripped state is by definition less trusted:

>  Although stripped state is usually generated and provided by the server, it is still possible to be incorrect on the receiving end. The stripped state events are not signed and could theoretically be modified, or outdated due to updates not being sent.

There's also this in the spec:
> Clients should only use stripped state events so long as they don’t have access to the proper state of the room. Once the state of the room is available, all stripped state should be discarded. In cases where the client has an archived state of the room (such as after being kicked) and the client is receiving stripped state for the room, such as from an invite or knock, then the stripped state should take precedence until fresh state can be acquired from a join.

While I don't think we need to discard it completely, we certainly can't mix things from stripped state and normal state.

Link to the relevant spec entry: https://spec.matrix.org/v1.2/client-server-api/#stripped-state

Reply:
Thanks for sharing that insight, I wasn't aware about this aspect.

> The list of joined and invited user ids is directly used to decide which members get the current room key.

But `UserId` isn't actually enough to share that information with, right? They are only an early indicator about the user_ids present in a room. So in the real-world scenario, we'd still be fetching the actual `Member`-Object, which indicates whether that information was given from stripped or not (or at least can tell). So, in reality this would still be okay-ish.

However
>  Once the state of the room is available, all stripped state should be discarded.

If that is an assumption made by spec, then even data extracted from the stripped state is probably necessary to be deleted. While we do delete the user_id from the other index, whenever we see a new membership event after, we don't necessarily remove them properly when we were told one in stripped_state that never actually happened. While practically unlikely to cause any problem, it would still leave us in an inconsistent state.

Okay. So, what is the proposed solution here then: second set of indizes for the stripped_states (that we purge any room info when we see it on the regular room) and do a fallback lookup on them the same way we handle the member info itself right now?

Reply:
> But `UserId` isn't actually enough to share that information with, right? They are only an early indicator about the user_ids present in a room. So in the real-world scenario, we'd still be fetching the actual `Member`-Object, which indicates whether that information was given from stripped or not (or at least can tell). So, in reality this would still be okay-ish.

Actually the `UserId` is enough, well as far as the member event is concerned. The joined/invited/left user ids lists are a store optimization. 

The state-store trait has a method to get the joined and invited user ids. How this is implemented is left to decide for each implementation. If they want to deserialize all member events and count that or if they have an index as a optimization doesn't matter. It does need to be correct and match to what the member events have.

If they do not match you'll risk having user ids for which you do not have the matching event which would lead to the client rendering one list of users while another list of users gets room keys.

> If that is an assumption made by spec, then even data extracted from the stripped state is probably necessary to be deleted. While we do delete the user_id from the other index, whenever we see a new membership event after, we don't necessarily remove them properly when we were told one in stripped_state that never actually happened. While practically unlikely to cause any problem, it would still leave us in an inconsistent state.

Precisely, this inconsistent state would lead to the above mentioned security vulnerability.

> Okay. So, what is the proposed solution here then: second set of indices for the stripped_states (that we purge any room info when we see it on the regular room) and do a fallback lookup on them the same way we handle the member info itself right now?

Correct, never mix events or info from stripped state with normal state. Rooms that are considered joined/left pick data from the normal buckets. Rooms that are considered invited/knocked pick data from the stripped state buckets.